### PR TITLE
Basic skeleton for life cycle code

### PR DESCRIPTION
**Pull request by alice-i-cecile**
This PR restructures quite a bit, both WRT inventory and crafting and when spawning bundles.

As a result, I'm putting this in ASAP to reduce merge conflicts.

### CODE REVIEW

File: emergence_lib/src/organisms/life_cycles.rs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    prelude::Component,
+    utils::{HashMap, HashSet},
+};
+
+use super::OrganismKind;
+
+/// A collection of [`LifeStage`]s for an organism of type `O`, connected by [`LifePath`]s
+#[derive(Component, Default)]
+pub struct LifeCycle<O: OrganismKind> {
+    /// The set of all possible life cycles for an organism of type `O`.
```
Comment (TimJentzsch) [inline emergence_lib/src/organisms/life_cycles.rs:13]: The struct is already called life cycle and we don't have recursion here
```suggestion
    /// The set of all possible life stages for an organism of type `O`.
```

**Comment by TimJentzsch** [comment]
I'll review this one before merging :)

File: emergence_lib/src/organisms/mod.rs [hunk 2]
```diff
@@ -3,24 +3,41 @@
...
+    life_cycles::LifeCycle,
     sessile::{fungi::FungiPlugin, plants::PlantsPlugin},
     units::UnitsPlugin,
 };
 
+pub mod life_cycles;
 pub mod sessile;
 pub mod units;
 
 /// All of the compone
```
Comment (TimJentzsch) [inline emergence_lib/src/organisms/mod.rs:15]: This looks like a typo
```suggestion
/// All of the components
```

File: emergence_lib/src/organisms/sessile/fungi.rs [hunk 3]
```diff
@@ -5,40 +5,63 @@ use bevy_ecs_tilemap::tiles::TilePos;
...
-    /// Fungi are structures.
-    structure_bundle: StructureBundle,
-    /// Position in the world
-    position: TilePos,
+pub struct LeucoBundle {
+    /// Acacias are organisms
+    organism_bundle: OrganismBundle<Leuco>,
+
+    /// Acacias are plants
+    plant: Fungi,
```
Comment (TimJentzsch) [inline emergence_lib/src/organisms/sessile/fungi.rs:29]: Fungis are not plants, this looks like a copy paste error

File: emergence_lib/src/organisms/sessile/mod.rs [hunk 4]
```diff
@@ -2,5 +2,37 @@
...
+    structures::{crafting::CraftingBundle, StructureBundle},
+};
+
 pub mod fungi;
 pub mod plants;
+
+/// Sessile organisms cannot move, and automatically process nutrients from their environment
+#[derive(Bundle)]
+pub struct SessileBundle {
+    /// Acacias are structures
```
Comment (TimJentzsch) [inline emergence_lib/src/organisms/sessile/mod.rs:19]:
```suggestion
    /// Sessile organisms are structures
```

File: emergence_lib/src/organisms/sessile/mod.rs [hunk 5]
```diff
@@ -2,5 +2,37 @@
...
+    pub structure_bundle: StructureBundle,
+
+    /// Plants can craft things
```
Comment (TimJentzsch) [inline emergence_lib/src/organisms/sessile/mod.rs:22]:
```suggestion
    /// Sessile organisms can craft things
```

File: emergence_lib/src/organisms/life_cycles.rs [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    pub life_stages: HashSet<O::LifeStage>,
+    /// A map from the current life stage to the potential ways it could transform.
+    pub life_paths: HashMap<O::LifeStage, LifePath>,
+}
+
+/// Paths that connect different life stages.
+///
+/// These are triggered when certain conditions are met for each organism,
+/// causing them to change form and function.
+pub struct LifePath;
```
Comment (TimJentzsch) [inline emergence_lib/src/organisms/life_cycles.rs:23]: Is this a placeholder for now?
Seems like the paths can't be distinguished at this time

**Review by TimJentzsch — COMMENTED** [review]
Hmm, I'm a bit torn.

Giving the individual plants and fungi their own marker components and making the lifecycles generic over the organism kind feels like the rusty/bevy thing to do.

On the other hand I'm a bit worried on how that will scale when we add tons of new organisms to the game.
I feel like then it will be harder to just load them via asset files and we'll need a lot more boilerplate code to create new stuff.

File: emergence_lib/src/organisms/life_cycles.rs (hunk 6 above)
Comment (alice-i-cecile) [reply emergence_lib/src/organisms/life_cycles.rs:23]: Yep, just a placeholder.

**Comment by alice-i-cecile** [comment]
> I feel like then it will be harder to just load them via asset files and we'll need a lot more boilerplate code to create new stuff.

I've cleaned up the design a bit in the latest commit. Overall, I think we're stuck with a pretty high complexity setup here. We ultimately need a finite state machine for the life cycles, and I am very reluctant to move away from enum types for the life stages due to how darn nice `match` is.

IMO we should reuse a finite state machine solution between life cycles and unit behavior. That might be `seldom_state`, but I need to check it out more.

**Review by TimJentzsch — APPROVED** [review]
I'm still not entirely sold on needing to define a bundle for every organism that we'll have in the game, but we can try it out for now

**Pull request merged**